src/core/OSTreeTUI.cpp

```diff
@@ -1,292 +1,395 @@
 #include <string>
-#include <unordered_map>
 #include <thread>
+#include <unordered_map>
 
-#include <fcntl.h>
-
-#include "commit.hpp"
+#include <ftxui/component/event.hpp>  // for Event, Event::ArrowDown, Event::ArrowUp, Event::End, Event::Home, Event::PageDown, Event::PageUp
 #include "ftxui/component/component.hpp"  // for Renderer, ResizableSplitBottom, ResizableSplitLeft, ResizableSplitRight, ResizableSplitTop
 #include "ftxui/component/component_base.hpp"      // for ComponentBase
 #include "ftxui/component/screen_interactive.hpp"  // for ScreenInteractive
-#include "ftxui/dom/elements.hpp"  // for Element, operator|, text, center, border
-
-#include "scroller.hpp"
-
-#include "footer.hpp"
-#include "manager.hpp"
+#include "ftxui/dom/elements.hpp"                  // for Element, operator|, text, center, border
 
 #include "clip.h"
 
 #include "../util/cpplibostree.hpp"
 
-std::vector<std::string> OSTreeTUI::parseVisibleCommitMap(cpplibostree::OSTreeRepo& repo,
-                            std::unordered_map<std::string, bool>& visibleBranches) {
-    
-	std::vector<std::string> visibleCommitViewMap{};
-	// get filtered commits
-	visibleCommitViewMap = {};
-	for (const auto& commitPair : repo.getCommitList()) {
-		if (visibleBranches[commitPair.second.branch]) {
-			visibleCommitViewMap.push_back(commitPair.first);
-		}
-	}
-	// sort by date
-	std::sort(visibleCommitViewMap.begin(), visibleCommitViewMap.end(), [&](const std::string& a, const std::string& b) {
-		return repo.getCommitList().at(a).timestamp
-			 > repo.getCommitList().at(b).timestamp;
-	});
-
-	return visibleCommitViewMap;
-}
+OSTreeTUI::OSTreeTUI(const std::string& repo, const std::vector<std::string>& startupBranches)
+    : ostreeRepo(repo), selectedCommit(0), screen(ftxui::ScreenInteractive::Fullscreen()) {
+    using namespace ftxui;
+
+    // set all branches as visible and define a branch color
+    for (const auto& branch : ostreeRepo.getBranches()) {
+        // if startupBranches are defined, set all as non-visible
+        visibleBranches[branch] = startupBranches.size() == 0 ? true : false;
+        std::hash<std::string> nameHash{};
+        branchColorMap[branch] = Color::Palette256((nameHash(branch) + 10) % 256);
+    }
+    // if startupBranches are defined, only set those visible
+    if (startupBranches.size() != 0) {
+        for (const auto& branch : startupBranches) {
+            visibleBranches[branch] = true;
+        }
+    }
+
+    // - UI ELEMENTS ---------- ----------
+
+    // COMMIT TREE
+    RefreshCommitComponents();
+
+    tree = Renderer([&] {
+        RefreshCommitComponents();
+        selectedCommit = std::min(selectedCommit, visibleCommitViewMap.size() - 1);
+        // TODO check for promotion & pass information if needed
+        if (inPromotionSelection && promotionBranch.size() != 0) {
+            std::unordered_map<std::string, Color> promotionBranchColorMap{};
+            for (auto& [str, col] : branchColorMap) {
```

Comment:
```suggestion
            for (const auto& [str, col] : branchColorMap) {
```

I know `auto` does it but I prefer `const` like `&`  to be explicit.

Reply:
I also prefer the explicit `const`, but seemed to have overlooked it here. Thanks :)